src/hooks/v1/contractReader/ContractReader.ts

```diff
@@ -31,72 +31,30 @@ export default function useContractReader<V>({
   callback?: (val?: V) => void
   valueDidChange?: (oldVal?: V, newVal?: V) => boolean
 }): V | undefined {
-  const [value, setValue] = useState<V | undefined>()
-
   const { contracts } = useContext(V1UserContext)
 
-  const _formatter = useCallback(
-    (val: any) => (formatter ? formatter(val) : val), // eslint-disable-line @typescript-eslint/no-explicit-any
-    [formatter],
-  )
   const _callback = useCallback(
     (val: any) => (callback ? callback(val) : val), // eslint-disable-line @typescript-eslint/no-explicit-any
     [callback],
   )
-  const _valueDidChange = useCallback(
-    (a?: any, b?: any) => (valueDidChange ? valueDidChange(a, b) : a !== b), // eslint-disable-line @typescript-eslint/no-explicit-any
-    [valueDidChange],
-  )
-
-  useDeepCompareEffectNoCheck(() => {
-    async function getValue() {
-      const readContract = contractToRead(contract, contracts)
-
-      if (!readContract || !functionName || args === null) return
-
-      try {
-        console.info('📚 [V1] Read >', functionName)
-
-        const result = await readContract[functionName](...(args ?? []))
-
-        const newValue = _formatter(result)
-
-        if (_valueDidChange(value, newValue)) {
-          console.info(
-            '📗 [V1] New >',
-            functionName,
-            { args },
-            { newValue },
-            { contract: readContract.address },
-          )
-          setValue(newValue)
-          _callback(newValue)
-        }
-      } catch (err) {
-        console.error(
-          '📕 [V1] Read error >',
-          functionName,
-          { args },
-          { err },
-          { contract: readContract.address },
-          contracts,
-        )
-
-        Sentry.captureException(err, {
-          tags: {
-            contract: typeof contract === 'string' ? contract : undefined,
-            contract_function: functionName,
-          },
-        })
 
-        setValue(_formatter(undefined))
-        _callback(_formatter(undefined))
-      }
-    }
+  const { value, refetchValue } = useContractReadValue({
+    contract,
+    contracts,
+    functionName,
+    args,
+    version: 'V1',
+    formatter,
+    valueDidChange,
+  })
 
-    getValue()
+  // Call the callback on contract read value changed
+  useDeepCompareEffectNoCheck(() => {
+    _callback(value)
+  }, [value])
 
-    const listener = () => getValue()
+  useDeepCompareEffectNoCheck(() => {
+    const listener = () => refetchValue()
```

Comment:
It is also possible we can pull this stuff out into a separate hook, but I will leave that for another day